File: 프로그래머스/2/84512. 모음 사전/모음 사전.swift
import Foundation

var stack = [String]()
var count = 0

func solution(_ word:String) -> Int {
    let word = word.map({ String($0) })
    let _ = DFS(word: word, words: ["A", "E", "I", "O", "U"])
    return count
}


func DFS(word: [String], words: [String]) -> Bool {
    if word == stack {
        return true
    }
    
    count += 1

    for w in words {
        if stack.count < 5 {
            stack.append(w)
            if DFS(word: word, words: words) {
                return true
            }
            stack.removeLast()
        }
    }
    
    return false

}

File: 프로그래머스/2/84512. 모음 사전/모음 사전.py
def solution(word):
    words = ['A', 'E', 'I', 'O', 'U']
    end = len(words)
    word = list(word)
    global idx, answer
    idx = answer = 0
    
    def DFS(arr):
        global idx, answer
        if len(arr) == end:
            return
        
        for i in range(end):
            arr.append(words[i])
            idx += 1
            if arr == word and answer == 0:  
                answer = idx
                return
            DFS(arr)
            arr.pop()
            
    DFS([])
    
    return answer
